### PR TITLE
mark auth param as sensitive for PHP 8.2

### PR DESCRIPTION
**Pull request by remicollet**
Also:
* refactor MINIT (split in each class sources file)
* use @generate-class-entries in stub files
* add RedisException and RedisClusterException in stub files

### CODE REVIEW

**Comment by remicollet** [comment]
To mark sensitive param, a simpler way is to simply add in the main MINIT something like:

```
#if PHP_VERSION_ID >= 80200
 	zend_mark_function_parameter_as_sensitive(&class_entry->function_table, "auth", 0);
#endif
```

Above is perhaps more suitable for a 5.3.x

So this PR is more for a future 5.4 release and it takes benefit of new 8.2 gen_stub features

**WARNING**: this mean developers have to use 8.2 to properly regenerate arginfo if needed

**Comment by remicollet** [comment]
Second commit simplify code but also fixes what looks like an old issue

Previously
```
    /* Base Exception class */
    exception_ce = zend_hash_str_find_ptr(CG(class_table), "RuntimeException", sizeof("RuntimeException") - 1);
```

But, in reflection
`    Class [ <internal:redis> class RedisClusterException extends Exception implements Throwable, Stringable ] {`

Now:
`    Class [ <internal:redis> class RedisClusterException extends RuntimeException implements Stringable, Throwable ] {`

So now  we really extend RuntimeException what seems to be the expected behavior (according to sources), but is this considered as a BC break ?

File: library.h [hunk 1]
```diff
@@ -35,6 +35,9 @@
     #define REDIS_VALUE_EXCEPTION(m) zend_value_error(m)
 #endif
 
+#if PHP_VERSION_ID < 80200
+#define zend_mark_function_parameter_as_sensitive(a,b,c)
+#endif
```
Comment (remicollet) [inline library.h]: this conditional part could be removed if https://github.com/php/php-src/pull/8931 is merged (so with 8.2.0beta1)

File: redis_cluster.h [hunk 2]
```diff
@@ -91,12 +91,11 @@
     } \
     resp_func(INTERNAL_FUNCTION_PARAM_PASSTHRU, c, ctx);
 
-extern const zend_function_entry *redis_cluster_get_methods(void);
-
-/* Create cluster context */
-zend_object *create_cluster_context(zend_class_entry *class_type);
+#endif
```
Comment (yatsukhnenko) [inline redis_cluster.h]: @remicollet CodeQL is unhappy with this `#endif` here.

File: redis_arginfo.h [hunk 3]
```diff
@@ -1433,3 +1433,30 @@ static const zend_function_entry class_Redis_methods[] = {
...
+};
+
+static zend_class_entry *register_class_Redis(void)
+{
+	zend_class_entry ce, *class_entry;
+
+	INIT_CLASS_ENTRY(ce, "Redis", class_Redis_methods);
+	class_entry = zend_register_internal_class_ex(&ce, NULL);
+
+	zend_mark_function_parameter_as_sensitive(&class_entry->function_table, "auth", 0);
```
Comment (yatsukhnenko) [inline redis_arginfo.h]: Do we also need to put `__construct` here?

File: library.h (hunk 1 above)
Comment (remicollet) [reply library.h]: Fixed

File: redis_arginfo.h (hunk 3 above)
Comment (remicollet) [reply redis_arginfo.h]: Good question ;)

`$options` is not really sensible, but may contains `auth` entry which is sensible.

So... don't know, as you prefer

**Review by yatsukhnenko — APPROVED** [review]

**Comment by TimWolla** [comment]
Note: You'll likely need to update this PR due to php/php-src@342e18f10518d3dd5381fcdd8e84da84d1a8540c which removes the `zend_mark_function_parameter_as_sensitive()` function and `@sensitive-param` in favor of a more generic solution.

**Comment by remicollet** [comment]
@TimWolla yes I'm aware of this, and waiting for **beta1** (hopefully with https://github.com/php/php-src/pull/8991) before updating this PR

**Comment by remicollet** [comment]
Rebased + arginfo regenerated using upcoming 8.2.0beta1

**Comment by remicollet** [comment]
@yatsukhnenko as beta1 is now released, this is ready for a merge

**Comment by yatsukhnenko** [comment]
@remicollet build fails with error `error: use of undeclared identifier 'ZEND_STR_SENSITIVEPARAMETER'` Is this a problem of GH Actions?

**Comment by remicollet** [comment]
> @remicollet build fails with error `error: use of undeclared identifier 'ZEND_STR_SENSITIVEPARAMETER'` Is this a problem of GH Actions?

Perhaps not yet on beta1...

> ✓ PHP Installed PHP 8.2.0-dev (332ac8ec2519e807c09c3f273c5a596ee4ab0c52)

This one is 11 days old...

**Comment by remicollet** [comment]
@yatsukhnenko CI is green :)

**Pull request merged**